### PR TITLE
[PLAT-488] [PLAT-516] Add Rich Kafka Consumer Health Monitoring

### PR DESCRIPTION
**Pull request by avatarneil**
## 📚 Purpose
This PR takes a stab at adding rich health monitoring to our Kafka Consumers. It does this by implementing a `healthStatus` context which can be passed around like such:
```mermaid

flowchart LR;
  ParentProcess("Parent Process (defines the mutable healthStatus)")

  StreamSource["Stream Source (updates the healthStatus based on consumer health)"]

  ApiServer>"Api Server (hosts a /health endpoint to be used by a container orchestrator (e.g. k8s))"]

  ParentProcess<-->StreamSource
  ParentProcess-->ApiServer
```

This enables the ApiServer to be aware of the running status of any StreamSources which are being utilized within it.

## 👌 Resolves:
PLAT-488

## 📦 Impacts:
See changesets

## ✅ PR Checklist
- [x] Write some unit tests for core functionality (noting that this really needs to be integration tested)

### CODE REVIEW

File: packages/mds-api-server/@types/index.ts [hunk 1]
```diff
@@ -71,3 +71,5 @@ export type ApiResponseLocalsVersion<V extends string> = ApiResponseLocals<'vers
 
 export type ApiVersionedResponse<V extends string, B = {}> = ApiResponse<B & { version: V }> &
   ApiResponseLocalsVersion<V>
+
+export type HealthStatus = { healthy: boolean; message?: string }
```
Comment (nessur) [inline packages/mds-api-server/@types/index.ts:75]: could we add a `freshness` timestamp? `updated_at` or `healthy_since` ?
Could be invaluable for debugging and validating health monitor activity.

File: packages/mds-stream/kafka/stream-consumer.ts [hunk 2]
```diff
@@ -28,9 +29,50 @@ export interface KafkaStreamConsumerOptions {
...
+  consumer.on(HEARTBEAT, async () => {
+    StreamLogger.debug('The Kafka consumer has sent a heartbeat to the coordinator')
+    lastHeartbeat = now()
+  })
+
+  /**
+   * If we haven't sent a heartbeat to the coordinator in over a minute, restart the consumer pod entirely
+   */
+  const livelinessHandler = async () => {
+    if (lastHeartbeat && now() - 60_000 > lastHeartbeat) {
```
Comment (nessur) [inline packages/mds-stream/kafka/stream-consumer.ts:58]: magic number should be declared by name, maybe default 60s + loaded from process.env

File: packages/mds-stream/kafka/stream-consumer.ts [hunk 3]
```diff
@@ -42,8 +84,12 @@ const createStreamConsumer = async (
 
     const kafka = new Kafka({ clientId, brokers })
     const consumer = kafka.consumer({ groupId })
+
     await consumer.connect()
     await Promise.all(asArray(topics).map(topic => consumer.subscribe({ topic, fromBeginning })))
+
+    registerLivelinessHandling(consumer, healthStatus)
```
Comment (nessur) [inline packages/mds-stream/kafka/stream-consumer.ts:91]: I really like that this matures the handling of all the consumer lifecycle events, even if not all of them result in a health-monitor status change.

File: packages/mds-api-server/@types/index.ts (hunk 1 above)
Comment (janedotx) [reply packages/mds-api-server/@types/index.ts:75]: ➕

**Comment by janedotx** [comment]
Looks good so far! I agree with all of @nessur 's feedback.

File: packages/mds-stream/kafka/stream-consumer.ts (hunk 2 above)
Comment (avatarneil) [reply packages/mds-stream/kafka/stream-consumer.ts:58]: 100%, I was planning on making this configurable too :)

File: packages/mds-api-server/@types/index.ts (hunk 1 above)
Comment (avatarneil) [reply packages/mds-api-server/@types/index.ts:75]: I think I'll do this, and also make it an object which enumerates any sources... e.g.
```
{
  sources: {
    someExampleKafkaConsumerName: { updated_at: 1234 },
    someOtherKafkaConsumer: { updated_at: 4321 }
  }
}
```

Then someday we can add other information about the sources (maybe their last heartbeat times, even)

Comment (avatarneil) [reply packages/mds-api-server/@types/index.ts]: Take a look at what I implemented between be25c93 and 50b5344
https://github.com/lacuna-tech/mds-core/compare/be25c93..50b5344

File: packages/mds-stream/kafka/stream-consumer.ts (hunk 2 above)
Comment (avatarneil) [reply packages/mds-stream/kafka/stream-consumer.ts]: Resolved in [0ec7375](https://github.com/lacuna-tech/mds-core/pull/991/commits/0ec73758ba3205d010e8c7704ae42dc93b69cfc6)

File: packages/mds-stream/kafka/stream-consumer.ts [hunk 4]
```diff
@@ -14,23 +14,98 @@
...
+import { cleanEnv, num } from 'envalid'
 import type { Consumer, EachMessagePayload } from 'kafkajs'
 import { Kafka } from 'kafkajs'
 import { StreamLogger } from '../logger'
 import type { StreamConsumer } from '../stream-interface'
 import { getKafkaBrokers } from './helpers'
 
+const { KAFKA_HEARTBEAT_TIMEOUT_MS } = cleanEnv(process.env, {
+  KAFKA_HEARTBEAT_TIMEOUT_MS: num({
+    default: 60_000,
```
Comment (drtyh2o) [inline packages/mds-stream/kafka/stream-consumer.ts:29]: `seconds(60)`?

**Review by drtyh2o — APPROVED** [review]
200 OK

File: packages/mds-stream/kafka/stream-consumer.ts (hunk 4 above)
Comment (avatarneil) [reply packages/mds-stream/kafka/stream-consumer.ts:29]: Good call!

**Review by nessur — APPROVED** [review]

**Review by janedotx — APPROVED** [review]

**Pull request merged**